### ISSUE TITLE
Issue: NAUTILUS (Qwen2.5-VL) checkpoint contains all-zero weights in many LLM layers

### ISSUE THREAD

**Issue by 1425562478** [H-EmbodVis/NAUTILUS#5]
# Issue: NAUTILUS (Qwen2.5-VL) checkpoint contains all-zero weights in many LLM layers

## Environment
- CUDA: 12.4
- PyTorch: 2.5.1+cu124
- Transformers: 4.51.3
- flash-attn: 2.7.3
- GPU: RTX 3090 (24GB)

## Problem Description
When loading the NAUTILUS Qwen checkpoint (`H-EmbodVis/Nautilus-qwen-instruct-7b`) using the provided `inference.py`, the model produces **gibberish/repetitive outputs** instead of coherent text. For example:
- `tre logo tre logo tre logo ...`
- `c c c c c c ...`
- `"But fish" "But fish" "But fish" ...`

## What I've Verified
1. ✅ Base model `Qwen/Qwen2.5-VL-7B-Instruct` works perfectly on the same environment.
2. ✅ Not a `flash-attn` issue — tested with both `flash_attention_2` and `eager`.
3. ✅ Not a `dtype` mismatch — pixel_values correctly cast to `bfloat16`.
4. ✅ Not a CUDA version issue — upgraded from cu118 to cu124, same result.
5. ✅ Downloaded files are complete — verified exact byte sizes against HuggingFace metadata:
   - `model-00001-of-00004.safetensors`: 4,809,451,776 bytes ✅
   - `model-00002-of-00004.safetensors`: 4,991,495,800 bytes ✅
   - `model-00003-of-00004.safetensors`: 4,991,495,888 bytes ✅
   - `model-00004-of-00004.safetensors`: 2,429,622,760 bytes ✅

## Key Finding: All-Zero Weights in LLM Layers
I inspected the safetensors weights and found that **many LLM layers contain entirely zero tensors**:

| Layer | # Zero Tensors | All Parameters Zero? |
|-------|---------------|---------------------|
| 3 | 3 | Partial |
| 4 | 3 | Partial |
| 5 | 10 | **Yes (all 12)** |
| 6 | 12 | **Yes (all 12)** |
| 7 | 2 | Partial |
| 8 | 8 | Partial |
| 9 | 12 | **Yes (all 12)** |
| 10 | 9 | Partial |
| 11 | 12 | **Yes (all 12)** |
| 12 | 3 | Partial |
| 13 | 8 | Partial |
| 14 | 8 | Partial |
| 25 | 3 | Partial |
| 26 | 3 | Partial |

**Total: 96 tensors are completely zero.**

Verification script:
```python
from safetensors.torch import load_file
sd = load_file("model-00002-of-00004.safetensors")
for k, v in sd.items():
    if k.startswith("model.layers.") and v.numel() > 0 and v.abs().max() == 0:
        print(f"ALL ZERO: {k}")
```

## Root Cause Analysis
1. The training script uses **LoRA** (`--use_lora True`) and **DeepSpeed ZeRO-2**.
2. The uploaded checkpoint appears to be a merged full-model checkpoint, but the merge/save process seems to have failed for many layers.
3. When I copy the zero-weight layers from the base model, the model can generate **partially coherent** text (no longer pure gibberish), but still exhibits repetition/quality degradation, suggesting even the non-zero layers may have been incorrectly merged.
4. The **nautilus-specific visual modules** (`nautilus_encoder`, `nautilus_dark_attn`, etc.) have valid non-zero weights.

## Request
Could you please:
1. Verify whether the uploaded checkpoint was correctly merged from the LoRA adapters?
2. Re-upload a corrected full-model checkpoint if the merge was incomplete?
3. Or alternatively, provide instructions on how to load the checkpoint properly (e.g., as a LoRA adapter on top of the base model)?

Thank you for your excellent work on NAUTILUS!

**Comment by Chengnotwang** [comment]
Hi, thank you for the detailed report and for the careful debugging.

We tried to reproduce this issue on our side by following the README environment setup, downloading the checkpoint directly from HuggingFace, and running inference with the provided script. Our reproduction environment is: CUDA 12.4, PyTorch 2.5.1, Python 3.12, and an H100 GPU. In our tests, the model works normally on both underwater and general-domain images, and we did not observe gibberish, repetitive outputs, or abnormal generation collapse.

Default Prompt: Describe this image.

For underwater scene:

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/bc219490-da2f-4bf6-a5b1-8a62701bc01d" />
The image shows an underwater scene with large rocks covered in algae and several sea urchins. A starfish is visible on one of the rocks.
<img width="2287" height="121" alt="Image" src="https://github.com/user-attachments/assets/354e540d-a7dd-42cf-b429-be0f8eba6062" />

For general scene:

<img width="286" height="176" alt="Image" src="https://github.com/user-attachments/assets/dc7b3191-8e4b-408d-bf21-b172292c9ab5" />
A woman in a tan jacket and blue pants stands on a wooden platform near a crosswalk, looking to her right. A black car is visible in the background, along with other traffic and pedestrians.
<img width="2258" height="120" alt="Image" src="https://github.com/user-attachments/assets/3a151405-77e3-4316-a290-16ab3136d558" />

infer.sh:
CUDA_VISIBLE_DEVICES=0 python scripts/inference.py --checkpoint "Nautilus-qwen-instruct-7b" --image "test.jpg" --prompt "Describe the image"

We also checked the checkpoint file you mentioned, `model-00002-of-00004.safetensors`, using the same type of zero-weight inspection script. On our downloaded checkpoint, we did **not** find any all-zero tensors in the LLM layers.
``` python
from safetensors.torch import load_file
sd = load_file("model-00002-of-00004.safetensors")
for k, v in sd.items():
    if k.startswith("model.layers.") and v.numel() > 0 and v.abs().max() == 0:
        print(f"ALL ZERO: {k}")
```

There was no output, indicating that no parameter satisfied the condition `v.numel() > 0 and v.abs().max() == 0`.
<img width="688" height="50" alt="Image" src="https://github.com/user-attachments/assets/c1f6c319-453a-498d-91f4-56dec7f26811" />

To be more rigorous, we further performed a more complete parameter-level inspection over the LLM layers, and the results also confirmed that there are no fully zeroed LLM layer parameters in our local checkpoint.

check_safetensors_zero_stats.py:
[check_safetensors_zero_stats.py](https://github.com/user-attachments/files/27116132/check_safetensors_zero_stats.py)

results:
[nautilus_7b_zero.json](https://github.com/user-attachments/files/27116160/nautilus_7b_zero.json)
One additional point worth clarifying is that seeing an all-zero `mask_token` does not indicate that the checkpoint is corrupted.
In DINOv2/ViT-style backbones, `mask_token` is a special learnable token used for masked image modeling or masked-token training. It is only involved when input patch tokens are explicitly replaced by the mask token during training, for example when a `masks` argument is provided.
During normal visual feature extraction/inference, no patch masking is applied. The image patches are embedded normally and passed through the vision transformer blocks directly. In this path, `mask_token` is not used to compute the visual features.
Therefore, even if `mask_token` is all zeros, it does not affect the extracted DINOv2 visual features during inference. This is different from LLM layer weights such as attention projections, MLP weights, or layer norms, which are actively used during every forward pass. An all-zero `mask_token` alone is expected/benign and should not cause repetitive or gibberish text generation.

We currently cannot reproduce the reported all-zero-weight issue with the checkpoint downloaded from HuggingFace.

From our side, the released checkpoint appears to be correctly saved and does not contain all-zero LLM layers. The NAUTILUS-specific visual modules and the LLM layers are both valid in our verification.

Thanks again for reporting this. We are happy to continue investigating if you can provide the checksum information or a minimal reproduction script.
